File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3071_minimum-operations-to-write-the-letter-y-on-a-grid/solutions/python3/solution.py
from typing import List

class Solution:
    def minimumOperationsToWriteY(self, grid: List[List[int]]) -> int:
        n = len(grid)
        mid = n // 2
        
        # Collect all Y positions
        y_positions = set()
        # Main diagonal and secondary diagonal
        for i in range(mid + 1):
            y_positions.add((i, i))
            y_positions.add((i, n - 1 - i))
        # Vertical line
        for i in range(mid, n):
            y_positions.add((i, mid))
        
        # Collect all non-Y positions
        non_y_positions = set()
        for r in range(n):
            for c in range(n):
                if (r, c) not in y_positions:
                    non_y_positions.add((r, c))
        
        # Count frequencies of values in Y and non-Y cells
        y_count = [0, 0, 0]  # Count of 0, 1, 2 in Y region
        non_y_count = [0, 0, 0]  # Count of 0, 1, 2 in non-Y region
        
        for r in range(n):
            for c in range(n):
                if (r, c) in y_positions:
                    y_count[grid[r][c]] += 1
                else:
                    non_y_count[grid[r][c]] += 1
        
        # Compute minimal operations required to make all Y one value and non-Y another value
        min_operations = float('inf')
        for y_val in range(3):
            for non_y_val in range(3):
                if y_val != non_y_val:
                    # Operations to make all Y cells y_val
                    y_ops = sum(y_count) - y_count[y_val]
                    # Operations to make all non-Y cells non_y_val
                    non_y_ops = sum(non_y_count) - non_y_count[non_y_val]
                    # Total operations
                    min_operations = min(min_operations, y_ops + non_y_ops)
        
        return min_operations

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3071_minimum-operations-to-write-the-letter-y-on-a-grid/solutions/java/solution.java
class Solution {
    public int minimumOperationsToWriteY(int[][] grid) {
        int n = grid.length;
        int mid = n / 2;
        int[] countsY = new int[3];
        int[] countsNotY = new int[3];
        
        // Count occurrences of values in Y and not-Y regions
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (c == r || c == n - r - 1 || (r >= mid && c == mid)) {
                    // This cell is part of the Y
                    countsY[grid[r][c]]++;
                } else {
                    // This cell is not part of the Y
                    countsNotY[grid[r][c]]++;
                }
            }
        }
        
        int minOperations = Integer.MAX_VALUE;
        
        // Evaluate all combinations of setting Y to one value and not-Y to another
        for (int yVal = 0; yVal < 3; yVal++) {
            for (int notYVal = 0; notYVal < 3; notYVal++) {
                if (yVal != notYVal) {
                    int operationsY = 0;
                    int operationsNotY = 0;
                    for (int val = 0; val < 3; val++) {
                        if (val != yVal) {
                            operationsY += countsY[val]; // Change all non-yVal in Y region to yVal
                        }
                        if (val != notYVal) {
                            operationsNotY += countsNotY[val]; // Change all non-notYVal in not-Y region to notYVal
                        }
                    }
                    minOperations = Math.min(minOperations, operationsY + operationsNotY);
                }
            }
        }
        
        return minOperations;
    }
}
